src/xmpp/xmpp-im/xmpp_tasks.cpp

```diff
@@ -891,19 +892,145 @@ void JT_Message::onGo()
 //----------------------------------------------------------------------------
 class JT_PushMessage::Private {
 public:
+    struct SubsData {
+        Subscriber *sbs;
+        int userData;
+    };
     EncryptionHandler *m_encryptionHandler;
+    QHash<QString, QList<SubsData> > subsData;
+    QList<SubsData> subsMData;
+
+    QString genKey(const QString &s1, const QString &s2)
+    {
+        return QString("%1&%2").arg(s1).arg(s2);
+    }
+
+    bool processXml(const QDomElement &root, QDomElement &e, Client *c, bool nested)
+    {
+        QString tagName = e.tagName();
+        QString xmlnsStr = e.attribute(QLatin1String("xmlns"));
+        QString key = genKey(tagName, xmlnsStr);
+        if (subsData.contains(key)) {
+            QList<SubsData> sdl = subsData.value(key);
+            foreach (const SubsData &sd, sdl) {
+                if (sd.sbs->xmlEvent(root, e, c, sd.userData, nested))
+                    return true;
+                if (e.tagName() != tagName || e.attribute(QLatin1String("xmlns")) != tagName)
+                    return false;
+            }
+        }
+        return false;
+    }
+
+    bool processMessage(Message &msg, bool nested) {
+        foreach (const SubsData &sd, subsMData) {
+            if (sd.sbs->messageEvent(msg, sd.userData, nested))
+                return true;
+        }
+        return false;
+    }
 };
 
+JT_PushMessage::Subscriber::~Subscriber()
+{
+}
+
+bool JT_PushMessage::Subscriber::xmlEvent(const QDomElement &root, QDomElement &e, Client *c, int userData, bool nested)
+{
+    Q_UNUSED(root)
+    Q_UNUSED(e)
+    Q_UNUSED(c)
+    Q_UNUSED(userData)
+    Q_UNUSED(nested)
+    return false;
+}
+
+bool JT_PushMessage::Subscriber::messageEvent(Message &msg, int userData, bool nested)
+{
+    Q_UNUSED(msg);
+    Q_UNUSED(userData);
+    Q_UNUSED(nested)
+    return false;
+}
+
 JT_PushMessage::JT_PushMessage(Task *parent, EncryptionHandler *encryptionHandler)
-:Task(parent)
+    : Task(parent)
+    , d(new Private)
 {
-    d = new Private;
     d->m_encryptionHandler = encryptionHandler;
 }
 
 JT_PushMessage::~JT_PushMessage()
 {
-    delete d;
+}
+
+void JT_PushMessage::subscribeXml(Subscriber *sbs, const QString &tagName, const QString &xmlnsStr, int userData)
+{
+    QString key = d->genKey(tagName, xmlnsStr);
+    QList<Private::SubsData> list = d->subsData.value(key);
+    foreach (const Private::SubsData &sd, list) {
+        if (sd.sbs == sbs)
+            return;
+    }
+    list.append({ sbs, userData });
+    d->subsData.insert(key, list);
+}
+
+void JT_PushMessage::unsubscribeXml(Subscriber *sbs, const QString &tagName, const QString &xmlnsStr)
+{
+    QString key = d->genKey(tagName, xmlnsStr);
+    QList<Private::SubsData> list = d->subsData.value(key);
```

Comment:
This code is plain wrong.
It remove entries from a temporary list, but the original list is not updated.
Use an iterator and loop it.value() entries to remove them.
Then if it's empty, you can easily remove it with d->subsData.erase(iterator)